src/sage/modules/diamond_cutting.py

```diff
@@ -259,19 +269,87 @@ def calculate_voronoi_cell(basis, radius=None, verbose=False):
         sage: V = calculate_voronoi_cell(matrix([[1, 0], [0, 1]]))
         sage: V.volume()
         1
+
+    TESTS:
+
+    Verify that :issue:`39507` is fixed::
+
+        sage: from sage.modules.free_module_integer import IntegerLattice
+        sage: v = vector(ZZ, [1,1,1,-1])
+        sage: L = IntegerLattice([v])
+        sage: print(v in L)
+        True
+        sage: print(L.closest_vector(v))
+        (1, 1, 1, -1)
+        sage: C = L.voronoi_cell()
+        sage: C.Hrepresentation()
+        (An inequality (-1, -1, -1, 1) x + 2 >= 0,
+         An inequality (1, 1, 1, -1) x + 2 >= 0)
+        sage: v = vector(ZZ, [1,1,-1])
+        sage: L = IntegerLattice([v])
+        sage: C = L.voronoi_cell()
+        sage: C.Hrepresentation()
+        (An inequality (-2, -2, 2) x + 3 >= 0, An inequality (2, 2, -2) x + 3 >= 0)
+        sage: C.Vrepresentation()
+        (A line in the direction (0, 1, 1),
+         A line in the direction (1, 0, 1),
+         A vertex at (0, 0, -3/2),
+         A vertex at (0, 0, 3/2))
+
+    Verify that :issue:`37086` is fixed::
+
+        sage: from sage.modules.free_module_integer import IntegerLattice
+        sage: l  = [7, 0, -1, -2, -1, -2, 7, -2, 0, 0, -2, 0, 7, -2, 0, -1, -2, -1, 7, 0 , -1, -1, 0, -2, 7]
+        sage: M = matrix(5, 5, l)
+        sage: C = IntegerLattice(M).voronoi_cell()
+        sage: C
+        A 5-dimensional polyhedron in QQ^5 defined as the
+        convex hull of 720 vertices
     """
     dim = basis.dimensions()
+    # LLL-reduce for efficiency.
+    basis = basis.LLL()
+    if radius is None:
+        # Convert the basis matrix to use RDF numbers for efficiency when we
+        # calculate the triangular matrix of the QR decomposition.
+        from sage.rings.real_double import RDF
+        tranposed_RDF_matrix = (basis.transpose()).change_ring(RDF)
+        R = tranposed_RDF_matrix.QR()[1]
+        # The length of the vector formed by the diagonal entries of R is an
+        # upper bound for twice the covering radius, so it is an upper bound
+        # on the length of the lattice vectors that need to be considered for
+        # diamond cutting. However, the value of the `radius` keyword is
+        # actually a squared length, so there is no square root in the
+        # following formula.
+        radius = sum(R[i,i]**2 for i in range(dim[0]))
+        # We then divide by 4 as we will divide the basis by 2 later on.
+        radius = radius / 4
```

Comment:
```suggestion
        radius = ceil(radius / 4)
```
The radius should be an integer (or rational number), not `RDF`.